core/src/main/java/io/questdb/griffin/SqlCompiler.java

```diff
@@ -1981,6 +1997,66 @@ private CompiledQuery executeWithRetries(
         throw SqlException.position(0).put("underlying cursor is extremely volatile");
     }
 
+    UpdateStatement generateUpdate(QueryModel updateQueryModel, SqlExecutionContext executionContext) throws SqlException {
+        // Update QueryModel structure is
+        // QueryModel with SET column expressions
+        // |-- QueryModel of select-virtual or select-choose of data selected for update
+        final QueryModel selectQueryModel = updateQueryModel.getNestedModel();
+
+        // First generate plan for nested SELECT QueryModel
+        final RecordCursorFactory updateToCursorFactory = codeGenerator.generate(selectQueryModel, executionContext);
+
+        // And then generate plan for UPDATE top level QueryModel
+        final IntList tableColumnTypes = selectQueryModel.getUpdateTableColumnTypes();
+        final ObjList<CharSequence> tableColumnNames = selectQueryModel.getUpdateTableColumnNames();
+        final int tableId = selectQueryModel.getTableId();
+        final long tableVersion = selectQueryModel.getTableVersion();
+        return generateUpdateStatement(
+                updateQueryModel,
+                tableColumnTypes,
+                tableColumnNames,
+                tableId,
+                tableVersion,
+                updateToCursorFactory
+        );
+    }
+
+    private static UpdateStatement generateUpdateStatement(
+            @Transient QueryModel updateQueryModel,
+            @Transient IntList tableColumnTypes,
+            @Transient ObjList<CharSequence> tableColumnNames,
+            int tableId,
+            long tableVersion,
+            RecordCursorFactory updateToCursorFactory
+    ) throws SqlException {
+        String tableName = Chars.toString(updateQueryModel.getTableName().token);
+        if (!updateToCursorFactory.supportTableRowId(tableName)) {
+            throw SqlException.$(updateQueryModel.getModelPosition(), "Only simple UPDATE statements without joins are supported");
+        }
+
+        // Check that updateDataFactoryMetadata match types of table to be updated exactly
+        RecordMetadata updateDataFactoryMetadata = updateToCursorFactory.getMetadata();
+
+        for (int i = 0, n = updateDataFactoryMetadata.getColumnCount(); i < n; i++) {
+            int virtualColumnType = updateDataFactoryMetadata.getColumnType(i);
+            CharSequence updateColumnName = updateDataFactoryMetadata.getColumnName(i);
+            int tableColumnIndex = tableColumnNames.indexOf(updateColumnName);
+            int tableColumnType = tableColumnTypes.get(tableColumnIndex);
+
+            if (virtualColumnType != tableColumnType) {
+                // get column position
+                ExpressionNode setRhs = updateQueryModel.getNestedModel().getColumns().getQuick(i).getAst();
+                int position = setRhs.position;
+                throw SqlException.inconvertibleTypes(position, virtualColumnType, "", tableColumnType, updateColumnName);
```

Comment:
when this thrown factory is not freed, looks like this code is untested with factories or functions that can leak memory